bpf/bpf_lxc.c

```diff
@@ -1943,6 +1966,80 @@ int tail_ipv4_policy(struct __ctx_buff *ctx)
 	return ret;
 }
 
+static __always_inline bool
+ipv4_to_endpoint_is_hairpin_flow(struct __ctx_buff *ctx, struct iphdr *ip4)
+{
+	__be16 client_port, backend_port, service_port;
+	struct ipv4_ct_tuple tuple = {};
+	struct lb4_backend *backend;
+	__be32 pod_ip, service_ip;
+	struct ct_entry *entry;
+	struct ct_map *map;
+	int err, l4_off;
+
+	/* Extract the tuple from the packet so we can freely access addrs and ports.
+	 * All values are in network byte order.
+	 */
+	l4_off = ETH_HLEN + ipv4_hdrlen(ip4);
+	err = ipv4_ct_extract_l4_ports(ctx, l4_off, CT_INGRESS, &tuple, NULL);
+	if (IS_ERR(err))
+		return false;
```

Comment:
Note that for the upstream patch, @ti-mo was relying on `lb4_extract_tuple()` to check whether the L4 protocol is a valid service protocol. Did you consider just copying the helper over from upstream, to keep things consistent?

Reply:
Right, so the first thing I did was trying with `extract_l4_port()` instead, which _does_ validate the L4 proto - but copies only one L4 port, when I need the two of them. So I switched to `ipv4_ct_extract_l4_ports()`, which indeed misses this.

Your suggestion makes sense, I'll just copy `lb4_extract_tuple()`. Thanks!